pkg/services/ngalert/notifier/channels/webex.go

```diff
@@ -0,0 +1,210 @@
+)
+
+const webexAPIURL = "https://webexapis.com/v1/messages"
+
+// WebexNotifier is responsible for sending alert notifications as webex messages.
+type WebexNotifier struct {
+	*Base
+	ns       notifications.WebhookSender
+	log      log.Logger
+	images   ImageStore
+	tmpl     *template.Template
+	orgID    int64
+	settings *webexSettings
+}
+
+// PLEASE do not touch these settings without taking a look at what we support as part of
+// https://github.com/prometheus/alertmanager/blob/main/notify/webex/webex.go
+// Currently, the Alerting team is unifying channels and (upstream) receivers - any discrepancy is detrimental to that.
+type webexSettings struct {
+	Message string `json:"message,omitempty" yaml:"message,omitempty"`
+	RoomID  string `json:"room_id,omitempty" yaml:"room_id,omitempty"`
+	APIURL  string `json:"api_url,omitempty" yaml:"api_url,omitempty"`
+	Token   string `json:"bot_token" yaml:"bot_token"`
+}
+
+func buildWebexSettings(factoryConfig FactoryConfig) (*webexSettings, error) {
+	settings := &webexSettings{}
+	err := factoryConfig.Config.unmarshalSettings(&settings)
+	if err != nil {
+		return settings, fmt.Errorf("failed to unmarshal settings: %w", err)
+	}
+
+	if settings.APIURL == "" {
+		settings.APIURL = webexAPIURL
+	}
+
+	if settings.Message == "" {
+		settings.Message = DefaultMessageEmbed
+	}
+
+	settings.Token = factoryConfig.DecryptFunc(context.Background(), factoryConfig.Config.SecureSettings, "bot_token", settings.Token)
+
+	u, err := url.Parse(settings.APIURL)
+	if err != nil {
+		return nil, fmt.Errorf("invalid URL %q", settings.APIURL)
+	}
+	settings.APIURL = u.String()
+
+	return settings, err
+}
+
+func WebexFactory(fc FactoryConfig) (NotificationChannel, error) {
+	notifier, err := buildWebexNotifier(fc)
+	if err != nil {
+		return nil, receiverInitError{
+			Reason: err.Error(),
+			Cfg:    *fc.Config,
+		}
+	}
+	return notifier, nil
+}
+
+// buildWebexSettings is the constructor for the Webex notifier.
+func buildWebexNotifier(factoryConfig FactoryConfig) (*WebexNotifier, error) {
+	settings, err := buildWebexSettings(factoryConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	logger := log.New("alerting.notifier.webex")
+
+	return &WebexNotifier{
+		Base: NewBase(&models.AlertNotification{
+			Uid:                   factoryConfig.Config.UID,
+			Name:                  factoryConfig.Config.Name,
+			Type:                  factoryConfig.Config.Type,
+			DisableResolveMessage: factoryConfig.Config.DisableResolveMessage,
+			Settings:              factoryConfig.Config.Settings,
+		}),
+		orgID:    factoryConfig.Config.OrgID,
+		log:      logger,
+		ns:       factoryConfig.NotificationService,
+		images:   factoryConfig.ImageStore,
+		tmpl:     factoryConfig.Template,
+		settings: settings,
+	}, nil
+}
+
+// WebexMessage defines the JSON object to send to Webex endpoints.
+type WebexMessage struct {
+	RoomID  string   `json:"roomId,omitempty"`
+	Message string   `json:"markdown"`
```

Comment:
Should be `markdown` or `message`?

```suggestion
	Message string   `json:"message"`
```

Reply:
It's weird, but it should markdown because that's the name that webex uses. They have 2 with are `message` and `markdown` and `markdown` is the markdown version of `message` 🤷 

I don't think we should expose this complexity to the user and instead allow to set the message (similar to other notifiers, that also call it "message' despite being something else in the payload)